UIforETW/PowerStatus.cpp

```diff
@@ -108,125 +108,179 @@ void CPowerStatusMonitor::SampleBatteryStat()
 	for (int deviceNum = 0; deviceNum < maxBatteries; deviceNum++)
 	{
 		SP_DEVICE_INTERFACE_DATA did = { sizeof(did) };
-
-		if (SetupDiEnumDeviceInterfaces(hdev,
+		const BOOL enumDeviceInterfacesResult =
+			SetupDiEnumDeviceInterfaces(hdev,
 										0,
 										&GUID_DEVCLASS_BATTERY,
 										deviceNum,
-										&did))
+										&did);
+		if (!enumDeviceInterfacesResult)
+		{
+			const DWORD lastErr = ::GetLastError( );
+			if (lastErr == ERROR_NO_MORE_ITEMS)
+			{
+				break; // Enumeration failed - perhaps we're out of items
+			}
+			continue;
+		}
+
+		DWORD bytesNeeded = 0;
+		const BOOL getDetailResult = SetupDiGetDeviceInterfaceDetail(hdev,
+										&did,
+										0,
+										0,
+										&bytesNeeded,
+										0);
+
+		if (getDetailResult)
+		{
+			throw std::logic_error("Expected SetupDiGetDeviceInterfaceDetail failure!");
+		}
+
+		if (GetLastError() != ERROR_INSUFFICIENT_BUFFER)
+		{
+			continue;
+		}
+
+
+		HLOCAL detailDataAlloc = LocalAlloc(LPTR, bytesNeeded);
+		if (detailDataAlloc == NULL)
 		{
-			DWORD bytesNeeded = 0;
-			SetupDiGetDeviceInterfaceDetail(hdev,
+			std::terminate();
+		}
+
+		auto pdidd =
+			reinterpret_cast<PSP_DEVICE_INTERFACE_DETAIL_DATA const>(detailDataAlloc);
+		pdidd->cbSize = sizeof(*pdidd);
+		
+		const BOOL batteryFound = SetupDiGetDeviceInterfaceDetail(hdev,
 											&did,
-											0,
-											0,
+											pdidd,
+											bytesNeeded,
 											&bytesNeeded,
 											0);
+		if (!batteryFound)
+		{
+			FreeLocalAlloc(detailDataAlloc);//use `detailDataAlloc` to suppress

```

Comment:
I don't understand why detailDataAlloc avoids any warning. The only warning I saw was a NULL pointer warning and that is avoided by checking for NULL (as you did) or by using std::vector.